[JSC] Add JIT support for Float16Array

Constellation opened the pull request:
#### f20f57564cef1db2c274f57754a5c90ec1af6a22
<pre>
[JSC] Add JIT support for Float16Array
<a href="https://bugs.webkit.org/show_bug.cgi?id=277672">https://bugs.webkit.org/show_bug.cgi?id=277672</a>
<a href="https://rdar.apple.com/133271787">rdar://133271787</a>

Reviewed by Keith Miller.

This patch implements JIT optimizations for Float16Array. Right now it is only enabled for
ARM64 since ARM v8.3 has half-float native CPU instructions. But we could extend it further
with newer instruction set.

Right now, we are not showing Float16 in B3 pipeline, and we are always convert it to Double via patchpoint.
In the future, we should add Float16 type into B3 as a first-citizen and support various Float16 instructions
in B3 opcodes. But for JS, it is not so important.

* JSTests/stress/array-message-passing.js:
* JSTests/stress/call-non-calleable-constructors-as-function.js:
* JSTests/stress/construct-typed-array-without-new-in-dfg-and-ftl.js:
* JSTests/stress/istypedarrayview-intrinsic.js:
* JSTests/stress/reflect-set.js:
(set get Uint8Array):
* JSTests/stress/resources/standalone-pre.js:
(isTypedArray):
* JSTests/stress/resources/typedarray-constructor-helper-functions.js:
* JSTests/stress/resources/typedarray-test-helper-functions.js:
* JSTests/stress/resources/v8-typedarray-helpers.js:
* JSTests/stress/typed-array-canonical-numeric-index-string.js:
* JSTests/stress/typed-array-sort-small.js:
(shouldBe):
* JSTests/stress/typedarray-access-monomorphic-neutered.js:
* JSTests/stress/typedarray-access-neutered.js:
* JSTests/stress/typedarray-configure-index.js:
* JSTests/stress/typedarray-getownproperty.js:
* JSTests/stress/typedarray-join.js:
* JSTests/stress/typedarray-view-string-properties-neutered.js:
* JSTests/stress/v8-float16-array-next.js: Added.
(f):
* JSTests/stress/v8-float16-array.js: Added.
(storeAndLoad):
(assertAlmostEquals):
(0.3.const.copied.new.Float32Array.source):
* Source/JavaScriptCore/assembler/ARM64Assembler.h:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.cpp:
(JSC::MacroAssemblerARM64::collectCPUFeatures):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::convertDoubleToFloat16):
(JSC::MacroAssemblerARM64::convertFloat16ToDouble):
(JSC::MacroAssemblerARM64::loadFloat16):
(JSC::MacroAssemblerARM64::moveZeroToFloat16):
(JSC::MacroAssemblerARM64::moveFloat16To16):
(JSC::MacroAssemblerARM64::move16ToFloat16):
(JSC::MacroAssemblerARM64::storeFloat16):
(JSC::MacroAssemblerARM64::supportsFloat16):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::supportsFloat16):
(JSC::MacroAssemblerARMv7::convertDoubleToFloat16):
(JSC::MacroAssemblerARMv7::convertFloat16ToDouble):
(JSC::MacroAssemblerARMv7::loadFloat16):
(JSC::MacroAssemblerARMv7::moveZeroToFloat16):
(JSC::MacroAssemblerARMv7::move16ToFloat16):
(JSC::MacroAssemblerARMv7::moveFloat16To16):
(JSC::MacroAssemblerARMv7::storeFloat16):
* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
(JSC::MacroAssemblerRISCV64::supportsFloat16):
(JSC::MacroAssemblerRISCV64::convertDoubleToFloat16):
(JSC::MacroAssemblerRISCV64::convertFloat16ToDouble):
(JSC::MacroAssemblerRISCV64::loadFloat16):
(JSC::MacroAssemblerRISCV64::moveZeroToFloat16):
(JSC::MacroAssemblerRISCV64::move16ToFloat16):
(JSC::MacroAssemblerRISCV64::moveFloat16To16):
(JSC::MacroAssemblerRISCV64::storeFloat16):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::convertDoubleToFloat16):
(JSC::MacroAssemblerX86_64::convertFloat16ToDouble):
(JSC::MacroAssemblerX86_64::loadFloat16):
(JSC::MacroAssemblerX86_64::moveZeroToFloat16):
(JSC::MacroAssemblerX86_64::move16ToFloat16):
(JSC::MacroAssemblerX86_64::moveFloat16To16):
(JSC::MacroAssemblerX86_64::storeFloat16):
(JSC::MacroAssemblerX86_64::supportsFloat16):
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::runWithDowncast):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::toTypedArrayType):
(JSC::forResizableTypedArray):
(JSC::needsScratchFPR):
(JSC::forInBy):
(JSC::isStateless):
(JSC::doesJSCalls):
(JSC::isMegamorphic):
(JSC::canBeViaGlobalProxy):
(JSC::InlineCacheCompiler::generateWithGuard):
(JSC::InlineCacheCompiler::generateAccessCase):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheArrayGetByVal):
(JSC::tryCacheArrayPutByVal):
(JSC::tryCacheArrayInByVal):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGCSEPhase.cpp:
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::shouldSpeculateFloat16Array):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileArithF16Round):
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValImpl):
(JSC::FTL::DFG::LowerDFGToB3::compilePutByVal):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOutput.cpp:
(JSC::FTL::Output::storeDoubleAsFloat16):
(JSC::FTL::Output::f16round):
(JSC::FTL::Output::loadFloat16AsDouble):
* Source/JavaScriptCore/ftl/FTLOutput.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::MathObject::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/PlatformHave.h:

Canonical link: <a href="https://commits.webkit.org/281895@main">https://commits.webkit.org/281895@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3dc332fd87b4cd4333e0ff31e4026baa084cf8b6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61388 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40748 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13972 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65340 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11936 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48426 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12211 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49587 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8287 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64285 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/37880 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53172 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30430 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34541 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10400 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10849 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/54491 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56349 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10700 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67071 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/60636 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5334 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10475 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56964 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5357 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53137 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57179 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4405 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/82404 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9230 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36552 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14379 "Found 1 new JSC stress test failure: wasm.yaml/wasm/fuzz/memory.js.wasm-eager-jettison (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37635 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38729 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37379 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Jarred-Sumner commented on Source/JavaScriptCore/assembler/MacroAssemblerARM64.cpp:
```diff
@@ -875,12 +875,21 @@ void MacroAssemblerARM64::collectCPUFeatures()
 
         s_lseCheckState = (hwcaps & HWCAP_ATOMICS) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
         s_jscvtCheckState = (hwcaps & HWCAP_JSCVT) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
+        s_float16CheckState = ((hwcaps & HWCAP_FPHP) && (hwcaps & HWCAP_ASIMDHP)) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
```
nice, i know that people use Bun on raspberry pi's on linux arm64 that probably do not have this instruction

Constellation replied:
Yeah, this is ARMv8.3 or later.

kmiller68 commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -44,12 +44,14 @@
 
 #define CHECK_DATASIZE_OF(datasize) static_assert(datasize == 32 || datasize == 64)
 #define CHECK_DATASIZE_OF_SIMD(datasize) static_assert(datasize == 32 || datasize == 64 || datasize == 128)
+#define CHECK_DATASIZE_OF_FLOAT16(datasize) static_assert(datasize == 16 || datasize == 32 || datasize == 64)
```
I don't understand this name? Shouldn't it be `CHECK_DATASIZE_OF_FLOATING_POINT`/`CHECK_DATASIZE_OF_FP` or something? It's not limited to `Float16` specifically. Although, I think the latter is more out of style for WebKit as a whole but better matches use in this file.

Constellation replied:
Changed.

kmiller68 commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -44,12 +44,14 @@
 #define MEMOPSIZE_OF(datasize) ((datasize == 8 || datasize == 128) ? MemOpSize_8_or_128 : (datasize == 16) ? MemOpSize_16 : (datasize == 32) ? MemOpSize_32 : MemOpSize_64)
 #define CHECK_DATASIZE() CHECK_DATASIZE_OF(datasize)
 #define CHECK_DATASIZE_SIMD() CHECK_DATASIZE_OF_SIMD(datasize)
+#define CHECK_DATASIZE_FLOAT16() CHECK_DATASIZE_OF_FLOAT16(datasize)
```
Ditto

Constellation replied:
Changed.

kmiller68 commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -3297,7 +3299,7 @@ class ARM64Assembler {
     ALWAYS_INLINE void stur(FPRegisterID rt, RegisterID rn, int simm)
     {
-        CHECK_DATASIZE_SIMD();
+        CHECK_FP_MEMOP_DATASIZE();
```
Does this work with data size == 8? `CHECK_FP_MEMOP_DATASIZE()` seems to accept that?

Constellation replied:
Yes, it works for 8bit too.

kmiller68 commented on Source/JavaScriptCore/assembler/ARM64Assembler.h:
```diff
@@ -2908,9 +2910,9 @@ class ARM64Assembler {
-        ASSERT(srcsize == 16 || srcsize == 32 || srcsize == 64);
-        ASSERT(dstsize != srcsize);
+        static_assert(dstsize == 16 || dstsize == 32 || dstsize == 64);
+        static_assert(srcsize == 16 || srcsize == 32 || srcsize == 64);
```
Nit could be `CHECK_DATASIZE_OF_FLOATING_POINT(dstsize)`/`CHECK_DATASIZE_OF_FLOATING_POINT(srcSize)`

Constellation replied:
Changed.

kmiller68 commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -274,6 +274,10 @@
 #define HAVE_LSE_INSTRUCTION 1
 #endif
 
+#if (CPU(ARM64E) && OS(DARWIN)) || (COMPILER(CLANG) && defined(__ARM_FEATURE_FP16_SCALAR_ARITHMETIC))
```
Probably worth accepting `CPU(ARM64) && PLATFORM(MAC)` too.

Constellation replied:
Right. Changed.

kmiller68 reviewed (approved):
r=me with comments

Constellation committed:
[JSC] Add JIT support for Float16Array
https://bugs.webkit.org/show_bug.cgi?id=277672
rdar://133271787

Reviewed by Keith Miller.

This patch implements JIT optimizations for Float16Array. Right now it is only enabled for
ARM64 since ARM v8.3 has half-float native CPU instructions. But we could extend it further
with newer instruction set.

Right now, we are not showing Float16 in B3 pipeline, and we are always convert it to Double via patchpoint.
In the future, we should add Float16 type into B3 as a first-citizen and support various Float16 instructions
in B3 opcodes. But for JS, it is not so important.

* JSTests/stress/array-message-passing.js:
* JSTests/stress/call-non-calleable-constructors-as-function.js:
* JSTests/stress/construct-typed-array-without-new-in-dfg-and-ftl.js:
* JSTests/stress/istypedarrayview-intrinsic.js:
* JSTests/stress/reflect-set.js:
(set get Uint8Array):
* JSTests/stress/resources/standalone-pre.js:
(isTypedArray):
* JSTests/stress/resources/typedarray-constructor-helper-functions.js:
* JSTests/stress/resources/typedarray-test-helper-functions.js:
* JSTests/stress/resources/v8-typedarray-helpers.js:
* JSTests/stress/typed-array-canonical-numeric-index-string.js:
* JSTests/stress/typed-array-sort-small.js:
(shouldBe):
* JSTests/stress/typedarray-access-monomorphic-neutered.js:
* JSTests/stress/typedarray-access-neutered.js:
* JSTests/stress/typedarray-configure-index.js:
* JSTests/stress/typedarray-getownproperty.js:
* JSTests/stress/typedarray-join.js:
* JSTests/stress/typedarray-view-string-properties-neutered.js:
* JSTests/stress/v8-float16-array-next.js: Added.
(f):
* JSTests/stress/v8-float16-array.js: Added.
(storeAndLoad):
(assertAlmostEquals):
(0.3.const.copied.new.Float32Array.source):
* Source/JavaScriptCore/assembler/ARM64Assembler.h:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.cpp:
(JSC::MacroAssemblerARM64::collectCPUFeatures):
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.h:
(JSC::MacroAssemblerARM64::convertDoubleToFloat16):
(JSC::MacroAssemblerARM64::convertFloat16ToDouble):
(JSC::MacroAssemblerARM64::loadFloat16):
(JSC::MacroAssemblerARM64::moveZeroToFloat16):
(JSC::MacroAssemblerARM64::moveFloat16To16):
(JSC::MacroAssemblerARM64::move16ToFloat16):
(JSC::MacroAssemblerARM64::storeFloat16):
(JSC::MacroAssemblerARM64::supportsFloat16):
* Source/JavaScriptCore/assembler/MacroAssemblerARMv7.h:
(JSC::MacroAssemblerARMv7::supportsFloat16):
(JSC::MacroAssemblerARMv7::convertDoubleToFloat16):
(JSC::MacroAssemblerARMv7::convertFloat16ToDouble):
(JSC::MacroAssemblerARMv7::loadFloat16):
(JSC::MacroAssemblerARMv7::moveZeroToFloat16):
(JSC::MacroAssemblerARMv7::move16ToFloat16):
(JSC::MacroAssemblerARMv7::moveFloat16To16):
(JSC::MacroAssemblerARMv7::storeFloat16):
* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
(JSC::MacroAssemblerRISCV64::supportsFloat16):
(JSC::MacroAssemblerRISCV64::convertDoubleToFloat16):
(JSC::MacroAssemblerRISCV64::convertFloat16ToDouble):
(JSC::MacroAssemblerRISCV64::loadFloat16):
(JSC::MacroAssemblerRISCV64::moveZeroToFloat16):
(JSC::MacroAssemblerRISCV64::move16ToFloat16):
(JSC::MacroAssemblerRISCV64::moveFloat16To16):
(JSC::MacroAssemblerRISCV64::storeFloat16):
* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::convertDoubleToFloat16):
(JSC::MacroAssemblerX86_64::convertFloat16ToDouble):
(JSC::MacroAssemblerX86_64::loadFloat16):
(JSC::MacroAssemblerX86_64::moveZeroToFloat16):
(JSC::MacroAssemblerX86_64::move16ToFloat16):
(JSC::MacroAssemblerX86_64::moveFloat16To16):
(JSC::MacroAssemblerX86_64::storeFloat16):
(JSC::MacroAssemblerX86_64::supportsFloat16):
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::runWithDowncast):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::toTypedArrayType):
(JSC::forResizableTypedArray):
(JSC::needsScratchFPR):
(JSC::forInBy):
(JSC::isStateless):
(JSC::doesJSCalls):
(JSC::isMegamorphic):
(JSC::canBeViaGlobalProxy):
(JSC::InlineCacheCompiler::generateWithGuard):
(JSC::InlineCacheCompiler::generateAccessCase):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheArrayGetByVal):
(JSC::tryCacheArrayPutByVal):
(JSC::tryCacheArrayInByVal):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGCSEPhase.cpp:
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::shouldSpeculateFloat16Array):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compileGetByVal):
(JSC::DFG::SpeculativeJIT::compile):
(JSC::DFG::SpeculativeJIT::compilePutByVal):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileArithF16Round):
(JSC::FTL::DFG::LowerDFGToB3::compileGetByValImpl):
(JSC::FTL::DFG::LowerDFGToB3::compilePutByVal):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/ftl/FTLOutput.cpp:
(JSC::FTL::Output::storeDoubleAsFloat16):
(JSC::FTL::Output::f16round):
(JSC::FTL::Output::loadFloat16AsDouble):
* Source/JavaScriptCore/ftl/FTLOutput.h:
* Source/JavaScriptCore/runtime/Intrinsic.h:
* Source/JavaScriptCore/runtime/JSDataViewPrototype.cpp:
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::MathObject::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/PlatformHave.h:

Canonical link: https://commits.webkit.org/281895@main
